File: internal/ast/ast.go
package ast

import "strings"

type Location struct {
	Filename string
	Line     int
	Column   int
}

type TypeKind string

const (
	TypeInvalid  TypeKind = ""
	TypeVoid     TypeKind = "void"
	TypeInt      TypeKind = "int"
	TypeFloat    TypeKind = "float"
	TypeBool     TypeKind = "bool"
	TypeString   TypeKind = "string"
	TypeNull     TypeKind = "null"
	TypeArray    TypeKind = "array"
	TypeFunction TypeKind = "func"
	TypeStruct   TypeKind = "struct"
)

type Type struct {
	Kind     TypeKind
	Name     string
	Nullable bool
	Elem     *Type
	Params   []Type
	Return   *Type
}

func (t Type) String() string {
	var out string
	switch t.Kind {
	case TypeArray:
		if t.Elem == nil {
			out = "array[?]"
		} else {
			out = "array[" + t.Elem.String() + "]"
		}
	case TypeFunction:
		parts := make([]string, 0, len(t.Params))
		for _, param := range t.Params {
			parts = append(parts, param.String())
		}
		returnType := "void"
		if t.Return != nil {
			returnType = t.Return.String()
		}
		out = "func(" + strings.Join(parts, ", ") + ") " + returnType
	case TypeStruct:
		out = t.Name
	default:
		out = string(t.Kind)
	}
	if t.Nullable {
		out += "?"
	}
	return out
}

func (t Type) Equal(other Type) bool {
	if t.Kind != other.Kind || t.Nullable != other.Nullable || t.Name != other.Name {
		return false
	}
	switch t.Kind {
	case TypeArray:
		return t.Elem != nil && other.Elem != nil && t.Elem.Equal(*other.Elem)
	case TypeFunction:
		if len(t.Params) != len(other.Params) {
			return false
		}
		for i := range t.Params {
			if !t.Params[i].Equal(other.Params[i]) {
				return false
			}
		}
		if t.Return == nil || other.Return == nil {
			return t.Return == other.Return
		}
		return t.Return.Equal(*other.Return)
	default:
		return true
	}
}

func Basic(kind TypeKind) Type {
	return Type{Kind: kind}
}

func ArrayOf(elem Type) Type {
	return Type{Kind: TypeArray, Elem: &elem}
}

func FuncType(params []Type, ret Type) Type {
	return Type{Kind: TypeFunction, Params: params, Return: &ret}
}

func Struct(name string) Type {
	return Type{Kind: TypeStruct, Name: name}
}

type Program struct {
	Statements []Statement
}

type Statement interface {
	Loc() Location
	statementNode()
}

type Expression interface {
	Loc() Location
	ExprType() Type
	SetExprType(Type)
	expressionNode()
}

type Param struct {
	Name string
	Type Type
}

type VarDecl struct {
	Location   Location
	Name       string
	Annotation Type
	Value      Expression
	Mutable    bool
}

func (s *VarDecl) Loc() Location  { return s.Location }
func (s *VarDecl) statementNode() {}

type Assignment struct {
	Location Location
	Target   Expression
	Value    Expression
}

func (s *Assignment) Loc() Location  { return s.Location }
func (s *Assignment) statementNode() {}

type Out struct {
	Location Location
	Value    Expression
}

func (s *Out) Loc() Location  { return s.Location }
func (s *Out) statementNode() {}

type TestDecl struct {
	Location Location
	Name     string
	Body     []Statement
}

func (s *TestDecl) Loc() Location  { return s.Location }
func (s *TestDecl) statementNode() {}

type Assert struct {
	Location Location
	Value    Expression
}

func (s *Assert) Loc() Location  { return s.Location }
func (s *Assert) statementNode() {}

type Import struct {
	Location Location
	Module   string
}

func (s *Import) Loc() Location  { return s.Location }
func (s *Import) statementNode() {}

type Export struct {
	Location Location
	Name     string
}

func (s *Export) Loc() Location  { return s.Location }
func (s *Export) statementNode() {}

type FuncDecl struct {
	Location   Location
	Name       string
	Receiver   string
	Params     []Param
	ReturnType Type
	Body       []Statement
}

func (s *FuncDecl) Loc() Location  { return s.Location }
func (s *FuncDecl) statementNode() {}

type StructField struct {
	Location Location
	Name     string
	Type     Type
}

type StructDecl struct {
	Location Location
	Name     string
	Fields   []StructField
}

func (s *StructDecl) Loc() Location  { return s.Location }
func (s *StructDecl) statementNode() {}

type Return struct {
	Location Location
	Value    Expression
}

func (s *Return) Loc() Location  { return s.Location }
func (s *Return) statementNode() {}

type If struct {
	Location Location
	Cond     Expression
	Then     []Statement
	Else     []Statement
}

func (s *If) Loc() Location  { return s.Location }
func (s *If) statementNode() {}

type While struct {
	Location Location
	Cond     Expression
	Body     []Statement
}

func (s *While) Loc() Location  { return s.Location }
func (s *While) statementNode() {}

type Repeat struct {
	Location Location
	Count    Expression
	Body     []Statement
}

func (s *Repeat) Loc() Location  { return s.Location }
func (s *Repeat) statementNode() {}

type For struct {
	Location Location
	Name     string
	Iterable Expression
	Body     []Statement
}

func (s *For) Loc() Location  { return s.Location }
func (s *For) statementNode() {}

type Break struct {
	Location Location
}

func (s *Break) Loc() Location  { return s.Location }
func (s *Break) statementNode() {}

type Continue struct {
	Location Location
}

func (s *Continue) Loc() Location  { return s.Location }
func (s *Continue) statementNode() {}

type ExprBase struct {
	Location Location
	Type     Type
}

func (e *ExprBase) Loc() Location      { return e.Location }
func (e *ExprBase) ExprType() Type     { return e.Type }
func (e *ExprBase) SetExprType(t Type) { e.Type = t }

type LiteralKind string

const (
	LiteralInt    LiteralKind = "int"
	LiteralFloat  LiteralKind = "float"
	LiteralBool   LiteralKind = "bool"
	LiteralString LiteralKind = "string"
	LiteralNull   LiteralKind = "null"
)

type Literal struct {
	ExprBase
	Kind  LiteralKind
	Value string
}

func (e *Literal) expressionNode() {}

type Name struct {
	ExprBase
	Identifier string
}

func (e *Name) expressionNode() {}

type Prefix struct {
	ExprBase
	Operator string
	Args     []Expression
}

func (e *Prefix) expressionNode() {}

type Call struct {
	ExprBase
	Callee   string
	Receiver Expression
	Method   string
	Args     []Expression
}

func (e *Call) expressionNode() {}

type ArrayLiteral struct {
	ExprBase
	Elements []Expression
}

func (e *ArrayLiteral) expressionNode() {}

type Index struct {
	ExprBase
	Target Expression
	Index  Expression
}

func (e *Index) expressionNode() {}

type FieldAccess struct {
	ExprBase
	Target Expression
	Field  string
}

func (e *FieldAccess) expressionNode() {}

File: compiler/ast/ast.cpp
#include "ast/ast.h"

#include <sstream>
#include <utility>

namespace walk::ast {

namespace {

std::string type_kind_name(TypeKind kind) {
    switch (kind) {
    case TypeKind::Invalid:
        return "";
    case TypeKind::Void:
        return "void";
    case TypeKind::Int:
        return "int";
    case TypeKind::Float:
        return "float";
    case TypeKind::Bool:
        return "bool";
    case TypeKind::String:
        return "string";
    case TypeKind::Null:
        return "null";
    case TypeKind::Array:
        return "array";
    case TypeKind::Function:
        return "func";
    case TypeKind::Struct:
    case TypeKind::Generic:
        return "";
    }
    return "";
}

}  // namespace

std::string Type::to_string() const {
    std::ostringstream output;
    switch (kind) {
    case TypeKind::Array:
        output << "array[";
        if (elem) {
            output << elem->to_string();
        } else {
            output << "?";
        }
        output << "]";
        break;
    case TypeKind::Function:
        output << "func(";
        for (std::size_t index = 0; index < params.size(); ++index) {
            if (index != 0) {
                output << ", ";
            }
            output << params[index].to_string();
        }
        output << ") ";
        if (return_type) {
            output << return_type->to_string();
        } else {
            output << "void";
        }
        break;
    case TypeKind::Struct:
    case TypeKind::Generic:
        output << name;
        break;
    default:
        output << type_kind_name(kind);
        break;
    }
    if (nullable) {
        output << "?";
    }
    return output.str();
}

Type basic(TypeKind kind) {
    Type type;
    type.kind = kind;
    return type;
}

Type array_of(Type elem) {
    Type type;
    type.kind = TypeKind::Array;
    type.elem = std::make_shared<Type>(std::move(elem));
    return type;
}

Type function_type(std::vector<Type> params, Type return_type) {
    Type type;
    type.kind = TypeKind::Function;
    type.params = std::move(params);
    type.return_type = std::make_shared<Type>(std::move(return_type));
    return type;
}

Type struct_type(std::string name) {
    Type type;
    type.kind = TypeKind::Struct;
    type.name = std::move(name);
    return type;
}

Type generic_type(std::string name) {
    Type type;
    type.kind = TypeKind::Generic;
    type.name = std::move(name);
    return type;
}

Node::Node(SourceRange node_range) : range(std::move(node_range)) {}
Node::~Node() = default;

Statement::Statement(StatementKind statement_kind, SourceRange statement_range)
    : Node(std::move(statement_range)), kind(statement_kind) {}
Statement::~Statement() = default;

Expression::Expression(ExpressionKind expression_kind, SourceRange expression_range)
    : Node(std::move(expression_range)), kind(expression_kind) {}
Expression::~Expression() = default;

VarDecl::VarDecl(SourceRange range) : Statement(StatementKind::VarDecl, std::move(range)) {}
Assignment::Assignment(SourceRange range) : Statement(StatementKind::Assignment, std::move(range)) {}
Out::Out(SourceRange range) : Statement(StatementKind::Out, std::move(range)) {}
Do::Do(SourceRange range) : Statement(StatementKind::Do, std::move(range)) {}
Defer::Defer(SourceRange range) : Statement(StatementKind::Defer, std::move(range)) {}
TestDecl::TestDecl(SourceRange range) : Statement(StatementKind::TestDecl, std::move(range)) {}
Assert::Assert(SourceRange range) : Statement(StatementKind::Assert, std::move(range)) {}
Import::Import(SourceRange range) : Statement(StatementKind::Import, std::move(range)) {}
Export::Export(SourceRange range) : Statement(StatementKind::Export, std::move(range)) {}
FuncDecl::FuncDecl(SourceRange range) : Statement(StatementKind::FuncDecl, std::move(range)) {}
StructDecl::StructDecl(SourceRange range) : Statement(StatementKind::StructDecl, std::move(range)) {}
Return::Return(SourceRange range) : Statement(StatementKind::Return, std::move(range)) {}
If::If(SourceRange range) : Statement(StatementKind::If, std::move(range)) {}
While::While(SourceRange range) : Statement(StatementKind::While, std::move(range)) {}
Repeat::Repeat(SourceRange range) : Statement(StatementKind::Repeat, std::move(range)) {}
For::For(SourceRange range) : Statement(StatementKind::For, std::move(range)) {}
Break::Break(SourceRange range) : Statement(StatementKind::Break, std::move(range)) {}
Continue::Continue(SourceRange range) : Statement(StatementKind::Continue, std::move(range)) {}

Literal::Literal(SourceRange range) : Expression(ExpressionKind::Literal, std::move(range)) {}
InterpolatedString::InterpolatedString(SourceRange range) : Expression(ExpressionKind::InterpolatedString, std::move(range)) {}
Name::Name(SourceRange range) : Expression(ExpressionKind::Name, std::move(range)) {}
Prefix::Prefix(SourceRange range) : Expression(ExpressionKind::Prefix, std::move(range)) {}
Call::Call(SourceRange range) : Expression(ExpressionKind::Call, std::move(range)) {}
Input::Input(SourceRange range) : Expression(ExpressionKind::Input, std::move(range)) {}
ArrayLiteral::ArrayLiteral(SourceRange range) : Expression(ExpressionKind::ArrayLiteral, std::move(range)) {}
Index::Index(SourceRange range) : Expression(ExpressionKind::Index, std::move(range)) {}
FieldAccess::FieldAccess(SourceRange range) : Expression(ExpressionKind::FieldAccess, std::move(range)) {}

}  // namespace walk::ast
